### PR TITLE
Cleanup ServiceLocator usages - Retrieve the services in a more specific fashion. 

### LINKED ISSUE #11198: Cleanup ServiceLocator for better performance
NuGet does a *lot* of custom things in the ServiceLocator. 

This task tracks cleaning up the unnecessary thread switches.

The gist is: 

* Avoid using methods that access the UI thread when request services exposed via MEF. Use free threaded calls, or ideally uses MEF composition. 
* Don't use custom methods to retrieve services from the service provider. Use the built in extensibility. 
* Avoid synchronous service lookup from async codepaths. This frequently leads to blocking the calling threadpool thread.

See tasks listed in this epic for more:

Internal issue [1389090](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1389090)

- [x] https://github.com/nuget/home/issues/11203
- [x] https://github.com/nuget/home/issues/11199

### LINKED ISSUE #11203: Review all sync ServiceLocator calls and move to async where possible
Note that this may *lead* some larger tasks. 

Use common sense. Fix the easy ones, create an issue for ones that can feasibly move to async codepaths. 

Note that a lot of the PMC ones cannot be made async, simple by virtue of PMC being mostly sync.

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11203
Fixes: https://github.com/NuGet/Home/issues/11198

Regression? Last working version:

## Description

The idea behind this long overdue refactoring is that we want every service retrieval to be intentional. 

You want a service from the service provider? Ask for it from the service provider.
You want a MEF service but you're not a MEF component (think PMC), ask for the service from MEF!

You want to actively block a thread on service retrieval? Explicitly do JTF.Run.

**open question**
A lot of this code removes hidden JTF.Runs. There is only 1 left, GetComponentModelService, which I debated removing, but I left because it has 12 usages. 
I'm happy to remove it if people think so.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs [hunk 1]
```diff
@@ -676,6 +676,8 @@ private async Task EnsureNuGetAndVsProjectAdapterCacheAsync()
                 {
                     try
                     {
+                        var itemsFactory = await ServiceLocator.GetGlobalServiceAsync<IVsEnumHierarchyItemsFactory, IVsEnumHierarchyItemsFactory>();
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: Are you making use of this? It's declared I don't see usage.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/MessageHelper.cs [hunk 2]
```diff
@@ -79,6 +79,11 @@ public static void ShowError(ErrorListProvider errorListProvider, TaskErrorCateg
             errorListProvider.ForceShowErrors();
         }
 
+        private static IServiceProvider GetServiceProvider()
+        {
+            return NuGetUIThreadHelper.JoinableTaskFactory.Run(() => ServiceLocator.GetServiceProviderAsync());
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/MessageHelper.cs:84]: Is it possible to cache this value?

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs [hunk 3]
```diff
@@ -27,49 +25,6 @@ public static void InitializePackageServiceProvider(IAsyncServiceProvider provid
...
-            var globalService = await GetGlobalServiceAsync<TService, TService>();
-            return globalService;
-        }
-
-        public static TInterface GetGlobalService<TService, TInterface>() where TInterface : class
-        {
-            return NuGetUIThreadHelper.JoinableTaskFactory.Run(GetGlobalServiceAsync<TService, TInterface>);
-        }
-
         public static async Task<TInterface> GetGlobalServiceAsync<TService, TInterface>() where TInterface : class
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs:34]: If I'm not mistaken, then this one resolves to MEF right?
Should we add small comment to this one to make clear for other developers?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs (hunk 1 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs]: oh I meant to remove that.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/MessageHelper.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/MessageHelper.cs:84]: Not easily, all these methods are static. 
Fortunately they are only invoked in rare failure scenarios, so the perf hit is lesser than the effort it'd take to refactor out how this error showing works.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ServiceLocator.cs:34]: No, It's not MEF. This hits the service provider.
 GetComponentModelServiceAsync is MEF.
 I added some docs.

**Review by erdembayar — APPROVED** [review]

**Pull request merged**